### PR TITLE
TrackJoint trajectory smoothing plugin

### PR DESCRIPTION
**Pull request by AndyZe**
Good news, PickNik has decided to open-source our jerk-limited trajectory smoothing algorithm, [TrackJoint](https://github.com/PickNikRobotics/trackjoint). TrackJoint is the product of about 2 man-years of development effort by several PhD's.

This is very experimental and I need to analyze the output to make sure it's working right. Please don't review yet -- the code is still littered with quick debug hacks.

This comes after trying fairly hard to get Ruckig trajectory smoothing working well. I went ahead and replaced the Ruckig smoothing algorithm with this PR but I guess it is up for debate whether we keep both plugins or not. The main obstacle for Ruckig is [documented in this issue](https://github.com/ros-planning/moveit2/issues/767).

TrackJoint runs slower than Ruckig but it is suited well for trajectories. Ruckig does not seem to work well with trajectories, at least for the free version.

TrackJoint is an iterative smoothing algorithm, so its runtime scales quickly based on how much smoothing is required. If only small corrections are needed, it runs very fast.

Requires this package be cloned into the workspace: `https://github.com/PickNikRobotics/trackjoint` (rolling branch)

You can easily test with `ros2 launch moveit_resources_panda_moveit_config demo.launch.py`. In panda_moveit_config/demo.launch.py add the TrackJoint plugin to the list of trajectory "request adapters". TrackJoint should be first in the list and AddTimeOptimalParameterization should be the next one. This ensures TOTG runs before TrackJoint.

`"request_adapters": """default_planner_request_adapters/AddTrackJointTrajectorySmoothing default_planner_request_adapters/AddTimeOptimalParameterization default_planner_request_adapters/ResolveConstraintFrames default_planner_request_adapters/FixWorkspaceBounds default_planner_request_adapters/FixStartStateBounds default_planner_request_adapters/FixStartStateCollision default_planner_request_adapters/FixStartStatePathConstraints""",`

For this robot:

- The TrackJoint timestep is set at 1 ms
- Joint vel limits are:  2.3925, 2.3925, 2.3925, 2.3925, 2.871, 2.871, 2.871 rad/s
- Joint accel limits are:  10, 10, 10, 10, 10, 10, 10
- Joint jerk limits are:  20, 20, 20, 20, 20, 20, 20

The active development branch is `rolling`.

slow-motion replay:

https://user-images.githubusercontent.com/11284393/141608524-b2ae153c-ce63-448f-96b5-b6f3ebc4f1ac.mp4

### CODE REVIEW

**Comment by AndyZe** [comment]
Example output. This is a graph of position for one joint. There is a little discontinuity when we move from one waypoint in the trajectory to the next, and that causes a jerk spike. I'm not quite sure how to fix this yet.
discontinuity_at_next_waypoint

**Comment by AndyZe** [comment]
I added a simple low-pass filter on the output to smooth the transitions between waypoints. In the future I would like to increase or decrease the amount of filtering until it achieves whatever jerk limit is required, but this should be a good start.

filtered_joint_states

**Comment by AndyZe** [comment]
Here's a mini data dump of plots for three joints:

- Input to TrackJoint from TOTG. This has a few sparse waypoints (26 of them, irregularly spaced in time)
- Output from TrackJoint, before the lowpass filter. 18,000 waypoints spaced at 1ms
- Final output after the lowpass filter

[data_for_three_joints.zip](https://github.com/ros-planning/moveit2/files/7614248/data_for_three_joints.zip)

Input to TrackJoint for J1:
j1_input_to_TJ

Output from TrackJoint:
j1_output_from_TJ

Final output from lowpass filter
final_output_j1_filtered
:

**Comment by AndyZe** [comment]
I found the bug that was causing discontinuities so I'm taking this out of WIP status now.  :+1: Luckily it was easy to fix.

I attached some data files of typical output.
[trackjoint_output.zip](https://github.com/ros-planning/moveit2/files/7780579/trackjoint_output.zip)

File: moveit_core/trajectory_processing/src/trackjoint_traj_smoothing.cpp [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+
+void TrackJointSmoothing::addTrackJointOutpointToRobotTrajectory(
+    const robot_trajectory::RobotTrajectory& reference_trajectory, const size_t num_dof,
+    const std::vector<int>& joint_group_indices, const std::vector<trackjoint::JointTrajectory>& trackjoint_output,
+    robot_trajectory::RobotTrajectory& new_trajectory)
+{
+  // We will overwrite this waypoint
+  moveit::core::RobotState waypoint = reference_trajectory.getWayPoint(0);
+
+  for (size_t waypoint_idx = 1; waypoint_idx < trackjoint_output.size(); ++waypoint_idx)
```
Comment (stephanie-eng) [inline moveit_core/trajectory_processing/src/trackjoint_traj_smoothing.cpp:189]:
```suggestion
  for (size_t waypoint_idx = 1; waypoint_idx < trackjoint_output.at(0).positions.size(); ++waypoint_idx)
```

File: moveit_core/trajectory_processing/src/trackjoint_traj_smoothing.cpp [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+  if (num_waypoints < 2)
+  {
+    RCLCPP_ERROR(LOGGER, "Trajectory does not have enough points to smooth with TrackJoint.");
+    return false;
+  }
+
+  robot_trajectory::RobotTrajectory outgoing_trajectory = reference_trajectory;
+  // Clear the waypoints
+  outgoing_trajectory.clear();
+  // ... but keep the first wayopint
```
Comment (stephanie-eng) [inline moveit_core/trajectory_processing/src/trackjoint_traj_smoothing.cpp:79]:
```suggestion
  // ... but keep the first waypoint
```

Comment (stephanie-eng) [inline moveit_core/trajectory_processing/src/trackjoint_traj_smoothing.cpp:79]:
```suggestion
  // ... but keep the first waypoint
```

File: moveit_core/trajectory_processing/src/trackjoint_traj_smoothing.cpp (hunk 1 above)
Comment (stephanie-eng) [inline moveit_core/trajectory_processing/src/trackjoint_traj_smoothing.cpp:189]:
```suggestion
  for (size_t waypoint_idx = 1; waypoint_idx < trackjoint_output.at(0).positions.size(); ++waypoint_idx)
```

**Comment by AndyZe** [comment]
Closing in favor of #1111

**Pull request closed without merging**